### PR TITLE
Specversion toggling

### LINKED ISSUE #54: CloudEvent only uses v1
## Expected Behavior
http_events.CloudEvent.from_http method currently only marshalls v1 events. This should probably be more dynamic in order to support multiple event versions. Perhaps we could restore specversion detection as seen in #36 or implement some other workaround. 

## Actual Behavior
Line 70 in https://github.com/cloudevents/sdk-python/blob/v1.0.0-dev/cloudevents/sdk/http_events.py
```python
        event = marshaller.NewDefaultHTTPMarshaller().FromRequest(
            v1.Event(), headers, data, data_unmarshaller)
```

### PR DESCRIPTION
**Pull request by cumason123**
Fixes #54

## Changes
from_http now marshalls with correct base event class depending on specversion. Created a converter is_binary function to givee from_http a hint as to whether given http headers is binary or not.

## One line description for the changelog
from_http now marshalls with correct base event class depending on specversion

- [*] Tests pass
- [*] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by cumason123** [comment]
I couldn't think of a unit test for this change because the v1 and v03 event classes functionally unmarshall data equivalently given any set of data. Idea's are welcome!

File: cloudevents/sdk/http_events.py [hunk 1]
```diff
@@ -64,8 +65,27 @@ def from_http(
...
-        event = marshaller.NewDefaultHTTPMarshaller().FromRequest(
-            v1.Event(), headers, data, data_unmarshaller
+        headers = {key.lower(): value for key, value in headers.items()}
+
+        content_type = headers.get("content-type", None)
+
+        specversion = "1.0"
+        marshall = marshaller.NewDefaultHTTPMarshaller()
+
+        for cnvrtr in marshall.http_converters:
```
Comment (grant) [inline cloudevents/sdk/http_events.py:75]: I understand you didn't introduce this change, but can you use full words `converter`?

File: cloudevents/sdk/http_events.py [hunk 2]
```diff
@@ -64,8 +65,27 @@ def from_http(
...
+            if cnvrtr.can_read(content_type):
+                if isinstance(cnvrtr, binary.BinaryHTTPCloudEventConverter):
+                    specversion = headers.get("ce-specversion", "1.0")
```
Comment (grant) [inline cloudevents/sdk/http_events.py:78]: Add a code comment as to what assumption this is making? (Assumption is fine)

File: cloudevents/sdk/http_events.py [hunk 3]
```diff
@@ -64,8 +65,27 @@ def from_http(
...
+                elif isinstance(cnvrtr, structured.JSONHTTPCloudEventConverter):
+                    raw_ce = json.loads(data)
+                    specversion = raw_ce.get("specversion", "1.0")
+                else:
+                    raise NotImplementedError
+
+        # Should this throw if given invalid event?
+        event_handlr = _obj_by_version.get(specversion, "1.0")
```
Comment (grant) [inline cloudevents/sdk/http_events.py:86]: `event_handler`

Comment (grant) [reply cloudevents/sdk/http_events.py:86]: Add a TODO instead of question.

```
# TODO: ...
```

File: cloudevents/sdk/http_events.py (hunk 1 above)
Comment (cumason123) [reply cloudevents/sdk/http_events.py:75]: LOL yes of course!

File: cloudevents/sdk/http_events.py (hunk 3 above)
Comment (cumason123) [reply cloudevents/sdk/http_events.py:86]: I simply resolved that TODO

**Comment by cumason123** [comment]
Added fix for #63 such that converters now do a simple check of whether ce-specversion is inside headers. I made this headers field optional to not break existing code, but it gives structured events the ability to recognize structured events in the absence of the application/cloudevents+json content-type

File: cloudevents/sdk/http_events.py [hunk 4]
```diff
@@ -64,8 +65,47 @@ def from_http(
...
-            v1.Event(), headers, data, data_unmarshaller
+        headers = {key.lower(): value for key, value in headers.items()}
+
+        content_type = headers.get("content-type", None)
+
+        specversion = None
+        marshall = marshaller.NewDefaultHTTPMarshaller()
+
+        for converter in marshall.http_converters:
+            # marshall atm has two converters, one for binary and another for
```
Comment (grayside) [inline cloudevents/sdk/http_events.py:76]: nit: avoid abbreviations in comments, makes it difficult for folks with less skill reading English.

File: cloudevents/sdk/http_events.py [hunk 5]
```diff
@@ -64,8 +65,47 @@ def from_http(
...
+            # structured. Binary converter will always return True, but the
```
[the remark this answers is not in the document]
Comment (grayside) [reply cloudevents/sdk/http_events.py:77]: The combination of something always returning true with the notion of can_read and checking instance types gives me the impression that something is unusual.

I would expect methods more around what the thing is: is it a Cloud Event? Is it set up for Structured mode or Binary mode? It seems like some of these evaluations have already been made given the isinstance calls below.

File: cloudevents/sdk/http_events.py [hunk 6]
```diff
@@ -64,8 +65,47 @@ def from_http(
...
+                    raw_ce = json.loads(data)
+
+                    specversion = raw_ce.get("specversion", None)
+                    # Breaking because while structured may or may not return
+                    # true on can_read, binary will always return true. Without
+                    # a break specversion for structured events will always be
+                    # set to None causing an error to be thrown.
+                    break
+                else:
+                    raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (grayside) [reply cloudevents/sdk/http_events.py:97]: This seems reasonable if further modes might be added in the future.

[the remark this answers is not in the document]
Comment (grant) [reply cloudevents/sdk/http_events.py:97]: Note: This was a multiline comment: `Comment on lines +81 to +97`.

The question isn't about NotImplementedError.

[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/http_events.py:97]: I'm unmarshalling event headers in order to determine the specversion so I know what event I'm dealing with while utilizing  legacy marshaller code. I'll try to modify the code to make it more verbose

File: cloudevents/sdk/http_events.py (hunk 4 above)
Comment (cumason123) [reply cloudevents/sdk/http_events.py:76]: I have a habit of doing that, good catch!

File: cloudevents/sdk/http_events.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/http_events.py:77]: I should remove that comment as it's no longer relevant. I added the headers argument to can_read for the converters so now they check headers to determine the cloudevent version. Previously binary simply always returned true which sort of worked given you would always test structured first. So the legacy code would always default to binary if it didn't know what was happening.

File: cloudevents/sdk/converters/binary.py [hunk 7]
```diff
@@ -25,8 +25,12 @@ class BinaryHTTPCloudEventConverter(base.Converter):
     TYPE = "binary"
     SUPPORTED_VERSIONS = [v03.Event, v1.Event]
 
-    def can_read(self, content_type: str) -> bool:
-        return True
+    def can_read(
```
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/converters/binary.py:28]: I'll restore the is_binary method

File: cloudevents/sdk/http_events.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/http_events.py:97]: But in a nutshell, after the converter tells me its able to read the event, I need to know which converter this was.

File: cloudevents/sdk/converters/binary.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/converters/binary.py:28]: http_events now has is_binary function which is separate from CloudEvent class

[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/converters/binary.py:28]: Moved is_binary and is_structured into cloudevents.sdk.converters.__init__

File: cloudevents/sdk/http_events.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/http_events.py]: Rewrote is_binary function in cloudevents.sdk.converters.__init__ to be more explicit with what is happening

File: cloudevents/sdk/http_events.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/http_events.py]: Completely refactored this logic

File: cloudevents/sdk/http_events.py [hunk 8]
```diff
@@ -64,8 +65,47 @@ def from_http(
...
-        event = marshaller.NewDefaultHTTPMarshaller().FromRequest(
-            v1.Event(), headers, data, data_unmarshaller
+        headers = {key.lower(): value for key, value in headers.items()}
+
+        content_type = headers.get("content-type", None)
+
+        specversion = None
+        marshall = marshaller.NewDefaultHTTPMarshaller()
+
+        for converter in marshall.http_converters:
```
[the remark this answers is not in the document]
Comment (cumason123) [reply cloudevents/sdk/http_events.py]: refactored

**Pull request merged**